components/clp-py-utils/clp_py_utils/sql_adapter.py

```diff
@@ -1,12 +1,82 @@
+import contextlib
+import functools
 import logging
+import time
 
 import mariadb
 import mysql.connector
+import sqlalchemy.pool as pool
 from mysql.connector import errorcode
+from sqlalchemy.dialects.mysql import mariadbconnector, mysqlconnector
 
 from clp_py_utils.clp_config import Database
 
 
+def exception_default_value(default):
+    """
+    This decorator wraps a function and returns the value specified by `default` whenever an
+    uncaught exception occurs in the function being wrapped. If no exception occurs this decorator
+    just forwards the return value of the function being wrapped.
+    """
+
+    def _exception_default_value(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            try:
+                return func(*args, **kwargs)
+            except:
+                return default
+
+        return wrapper
+
+    return _exception_default_value
+
+
+class DummyCloseableObject:
+    def close(self):
+        pass
+
+
+class ConnectionPoolWrapper:
+    """
+    This class implements a wrapper around an sqlalchemy connection pool in order to simplify error
+    handling, and prevent log spew when the database is down.
+    """
+
+    def __init__(self, pool: pool.QueuePool, logger: logging.Logger):
+        self.pool = pool
+        self.last_exception_logging_time = None
+        self.error_reporting_interval = 60  # one minute
+        self.logger = logger
+
+    def connect(self):
+        """
+        Check out a connection from the connection pool and return it. On error return a dummy
+        object which implements close, and periodically log the error so long as a connection error
+        has not occurred within the past interval.
+        """
+        try:
+            return self.pool.connect()
+        except:
+            current_time = time.time()
+            if (
+                self.last_exception_logging_time is None
+                or current_time - self.last_exception_logging_time > self.error_reporting_interval
+            ):
+                self.last_exception_logging_time = current_time
+                self.logger.exception(
```

Comment:
Just have one concern. After current connection fail, it looks like that the scheduler will keep retrying with the scheduelr polling interval. But on a database disconnection would it make more sense to wait for a longer time?

If we decide to do so, perhaps we can add a new field, say reconnection_interval under the scheduler config and let create_connection_pool take it in as a parameter?

But since the database is down anyway, retrying at a very short interval wouldn't cause any problem or increase the database traffice. Assuming Kirk will still review this PR, I would like to also get his opinion.

Reply:
Right, that makes sense to me, and the change is pretty easy to implement, so I think we should do it. I suspect that Kirk will have comments that touch some similar parts of the code though, so I'll wait for @kirkrodrigues 's review to make this change.

Reply:
I prefer to limit the number of knobs a user can tune since that increases our maintenance overhead (docs + semantics + harder to remove options later).

Is the concern about increased CPU usage while the database is down? I figure the database should be down for a short amount of time if at all (if it's down for longer, CLP should probably go down with it), but perhaps you've seen a different usage scenario.

Reply:
Let's address this in another PR to avoid adding more functionality than is immediately necessary.